### ISSUE TITLE
Refactor AST Processing for tags, links and tasks

### ISSUE THREAD

**Issue by mohamedsalem401** [datopian/markdowndb#65]
We will extract the conversion of source to AST from the `ExtractlinksFromBody` to `processAST` function
 ```
 // This is parsing the AST for the link extraction
  const processor = unified()
    .use(markdown)
    .use([
      gfm,
      [
        remarkWikiLink,
        { pathFormat: "obsidian-short", permalinks: options?.permalinks },
      ],
      ...userRemarkPlugins,
    ]);

  const ast = processor.parse(source);
``` 

then instead of 
```
  const bodyTags = extractTagsFromBody(source);
  const links = extractWikiLinks(options?.from || "", source, {
    permalinks: options?.permalinks,
  });
```

we will do:
```
  const AST = processAST(src)
  const bodyTags = extractTagsFromASt(AST);
  const links = extractWikiLinks(options?.from || "", AST, {
    permalinks: options?.permalinks,
  });
  // TODO
  const tasks = extractTasks(AST)
```

**Comment by mohamedsalem401** [comment]
@rufuspollock,

We can optimize our process even more by iterating through all the nodes of the AST in a single pass. Currently, we perform the following steps:

**For Links:**
```plaintext
links: source => AST => select all link nodes, select all wiki nodes, select all image nodes
```

**For Tags:**
```plaintext
tags: source => AST => select all nodes => get text nodes with a tag e.g., `#Tag`
```

To improve efficiency, we can consider the following approach:

```plaintext
AST => select all nodes =>
  if (node is a wikilink) => handle wiki links
  if (node is an image) => handle image
  if (node is a tag) => handle tag
  if (node is a task) => handle task

return {tags, links, tasks}
```

This way, we can consolidate the node selection and handle each node type in a more streamlined manner, potentially saving processing time.

**Comment by rufuspollock** [comment]
@mohamedsalem401 i wouldn't think *too* much about optimization at this point - more about the simplicity of the code. (i know i wasn't completely clear on this). so i think the first example is correct. The overall signature for our functions extracting info will be something like:

```
const listOfXXX = extractXxx(ast, options)
```

**Comment by rufuspollock** [comment]
FIXED in #66
